### ISSUE TITLE
InvalidArgumentError (see above for traceback): indices[0] = 1 is not in [0, 1)

### ISSUE THREAD

**Issue by donigian** [kundajelab/deeplift#52]
I'm getting the following error & am not sure how to resolve. 

```
HEADS UP! integrated_grads_5 and integrated_grads_10 take 5x and 10x longer to run respectively
Consider leaving them out to get faster results
Computing scores for: revealcancel
	Computing scores for task: 0
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
~/anaconda3/envs/yourenvname/lib/python3.5/site-packages/tensorflow/python/client/session.py in _do_call(self, fn, *args)
   1349     try:
-> 1350       return fn(*args)
   1351     except errors.OpError as e:
...
...
...
InvalidArgumentError (see above for traceback): indices[0] = 1 is not in [0, 1)
	 [[Node: ScatterUpdate_14 = ScatterUpdate[T=DT_FLOAT, Tindices=DT_INT32, _class=["loc:@Variable_21"], use_locking=true, _device="/job:localhost/replica:0/task:0/device:CPU:0"](Assign_63, ScatterUpdate_14/indices, ScatterUpdate_14/updates)]]
```

Here's the code snippet, error occurs when calling `score_func`.
```
from  collections  import OrderedDict
method_to_task_to_scores = OrderedDict()
print("HEADS UP! integrated_grads_5 and integrated_grads_10 take 5x and 10x longer to run respectively")
print("Consider leaving them out to get faster results")
for method_name, score_func in [
                               ('revealcancel', revealcancel_func)
                               ('guided_backprop_masked', guided_backprop_func_masked),
                               ('guided_backprop_times_inp', guided_backprop_times_inp_func),
                               ('simonyan_masked', simonyan_func_masked), 
                               ('grad_times_inp', grad_times_inp_func),
                               ('integrated_grads_5', integrated_grads_5),
                               ('integrated_grads_10', integrated_grads_10)
]:
    print("Computing scores for:",method_name)
    method_to_task_to_scores[method_name] = {}
    for task_idx in range(10):
        print("\tComputing scores for task: "+str(task_idx))

        scores = np.array(score_func(
                    task_idx=task_idx,
                    input_data_list=[X_test.values],
                    input_references_list=[np.zeros_like(X_test.values)],
                    batch_size=1000,
                    progress_update=None))
        method_to_task_to_scores[method_name][task_idx] = scores
```

Environment:

- Using `dev-tf` branch with TF v1.5.0, Keras 1.2.0 & DeepLIFT v0.5.1-tensorflow. MNIST notebook works using this environment.

- I'd like to explain a NN model I built using Keras front-end & TF backend w/ 2-dim tabular data (binary logistic regression).

- Running the MNIST example notebook with 2-dim X_train & X_test (both data frames but using `.values` to convert to ndarray ). 

- `difference in predictions: 0.0`

- `target_layer_idx=-2` for binary logistic regression

Could this problem be due to MNIST being a multi-class vs my task being a binary regression? If so, It's not clear what to modify in MNIST notebook to indicate binary logistic regression.

**Comment by AvantiShri** [comment]
If you have only one task, then rather than iterating over `for task_idx in range(10):`, just always have `task_idx=0`. Let me know if this addresses the issue.

**Comment by donigian** [comment]
Thanks @AvantiShri, that worked. Thanks again for not only making this awesome contribution, but providing support.

Before we close this ticket out, can you please elaborate on:

1. In what unit is the attribution scores/value of each algorithm (specifically DeepLIFT, IG-5)? For instance, I know that SHAP attribution values are in log-order. I'm trying to understand how to compare attribution scores across various algorithms. Essentially is a .5 from DeepLIFT equivalent to .5 from IG etc.

2. How does the implementation of IG in DeepLIFT compare with (DeepExplain)[https://github.com/marcoancona/DeepExplain] or (this)[https://github.com/ankurtaly/Integrated-Gradients] or (that)[https://github.com/hiranumn/IntegratedGradientsTF]. Out of curiosity, trying to figure out how you have validated your implementation of IG. I plan to run some benchmarks, just figured I ask.

3. During your talk @NVIDIA GTC, you mentioned an additional benefit of using DeepLIFT besides feature attribution for an instance, is using DeepLIFT to find patterns. Can you point me to some examples in a notebook where you found patterns. I believe you mentioned the genomics example maybe...

**Comment by AvantiShri** [comment]
Hi @donigian,

Glad the fix worked, and thanks for your kind words!

(1) The DeepLIFT attribution scores sum up linearly; the sum of attribution scores equals the difference from reference. A 0.5 from DeepLIFT is analogous to a 0.5 from IG.
(2) I tested my IG implementation using this unit test: https://github.com/kundajelab/deeplift/blob/671ee67a03bd5bebf4c405af59eec45d3ca2a288/tests/blobs/test_integrated_grads.py. I haven't looked at other implementations of IG, but I expect they should all produce similar results within numerical precision. The IG algorithm is straightforward to implement - just average the gradient at linearly interpolated points, then multiply by the difference-from-reference.
3. That was actually a different algorithm - TF-MoDISco. I have released code for it but no preprint since it's still in development. Here is the repo: https://github.com/kundajelab/tfmodisco
